### DISCUSSION TITLE
API testing against custom external modules

### DISCUSSION THREAD

**Discussion by nicpitsch** [oqtane/oqtane.framework#946] [Q&A] [answered]
Right now there seems to be no possibility to do API testing against custom external modules. My questions:

- Swagger isn't (yet) available for custom modules according to [issue218](https://github.com/oqtane/oqtane.framework/issues/218#issuecomment-589964887). Will it be realized in the near future?
- I tried Postman but with no success. It is possible to login with *api/User/login*, but I don't know how to send authorization info using subsequent api calls. Is there an instruction available how to do with Oqtane and Postman?
- Are there other ways to send API calls without a UI?

Thanks in advance.

**Comment by hishamco** [comment, marked as answer]
You can use Swagger by adding `UseSwagger` flag in `appsettings.json` and set it to `true`

**Reply by nicpitsch** [reply]
Thanks @hishamco for the information. I didn't know the *UseSwagger* flag. Without this flag Swagger is enabled too (by default), that's my starting point. The question isn't how to enable/disable Swagger. It's about it's functionality regarding **custom external modules**. Swagger is fully functional with the Oqtane framework. If you try out e.g. *GET /{alias}/api/Alias* it will return a Response body containing JSON.

Swagger lists the routes of **custom external modules**. But trying them out always returns a Response body containing the error message *An unhandled exception has occurred. See browser dev tools for details.*:

![2020-11-25_05-02-30](https://user-images.githubusercontent.com/1652835/100181958-bdaed800-2edb-11eb-96c3-f0e39e34389b.png)

**Reply by hishamco** [reply]
> Swagger isn't (yet) available for custom modules

Seems I miss understood your statement ;) I will have a look to this

**Reply by hishamco** [reply]
> Are there other ways to send API calls without a UI?

You could use Oqtane as Headless, please refer to https://github.com/oqtane/oqtane.framework/discussions/946#discussioncomment-138260
